src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -72,50 +73,75 @@ internal int FindMetricAggregators(ReadOnlySpan<KeyValuePair<string, object>> ta
 
-            lock (this.lockKeyValue2MetricAggs)
+            string[] seqKey = null;
+
+            // GetOrAdd by TagKey at 1st Level of 2-level dictionary structure.
+            // Get back a Dictionary of [ Values x Metrics[] ].
+            if (!this.keyValue2MetricAggs.TryGetValue(tagKey, out var value2metrics))
             {
-                string[] seqKey = null;
+                // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
+                seqKey = new string[len];
+                tagKey.CopyTo(seqKey, 0);
 
-                // GetOrAdd by TagKey at 1st Level of 2-level dictionary structure.
-                // Get back a Dictionary of [ Values x Metrics[] ].
-                if (!this.keyValue2MetricAggs.TryGetValue(tagKey, out var value2metrics))
+                value2metrics = new ConcurrentDictionary<object[], int>(new ObjectArrayEqualityComparer());
+                if (!this.keyValue2MetricAggs.TryAdd(seqKey, value2metrics))
                 {
-                    // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-                    seqKey = new string[len];
-                    tagKey.CopyTo(seqKey, 0);
-
-                    value2metrics = new Dictionary<object[], int>(new ObjectArrayEqualityComparer());
-                    this.keyValue2MetricAggs.Add(seqKey, value2metrics);
+                    // Some other thread added value2metrics, take it.
+                    // The one created by this thread is wasted.
+                    this.keyValue2MetricAggs.TryGetValue(tagKey, out value2metrics);
                 }
+            }
 
-                // GetOrAdd by TagValue at 2st Level of 2-level dictionary structure.
-                // Get back Metrics[].
-                if (!value2metrics.TryGetValue(tagValue, out aggregatorIndex))
+            // GetOrAdd by TagValue at 2st Level of 2-level dictionary structure.
+            // Get back Metrics[].
+            if (!value2metrics.TryGetValue(tagValue, out aggregatorIndex))
+            {
+                this.metricPointIndex++;
+                aggregatorIndex = this.metricPointIndex;
+                if (aggregatorIndex >= MaxMetricPoints)
                 {
-                    this.metricPointIndex++;
-                    aggregatorIndex = this.metricPointIndex;
-                    if (aggregatorIndex >= MaxMetricPoints)
-                    {
-                        // sorry! out of data points.
-                        return -1;
-                    }
+                    // sorry! out of data points.
+                    // TODO: This doesn't account for "wasted space"
+                    // in the array. Need to reclaim them when we
+                    // really need. i.e inside this if block.
+                    // this can take explicit lock, as this
+                    // condition is not the common code path.
+                    return -1;
```

Comment:
Tricky to do with the conconcurrentdictionary, but if we do the hashtable thing I mentioned above we could store the -1 so we don't have to re-evaluate this logic each time FindMetricAggregators is called (once we fill up)?